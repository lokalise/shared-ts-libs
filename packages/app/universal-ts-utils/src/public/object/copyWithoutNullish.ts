import type { RecordKeyType } from '../../internal/types.js'

type Output<T extends Record<RecordKeyType, unknown>> = Pick<
  T,
  {
    [Prop in keyof T]: T[Prop] extends null | undefined ? never : Prop
  }[keyof T]
>

/**
 * Creates a shallow copy of an object, excluding properties with `null` or `undefined` values.
 *
 * This function iterates over an object's own enumerable properties and creates a new
 * object that excludes properties with `null` or `undefined` values.
 *
 * @param {Record} object - The source object from which to copy properties.
 * @returns {Record} A new object containing only the properties from the source object
 *    that do not have `null` or `undefined` values.
 */
export const copyWithoutNullish = <T extends Record<RecordKeyType, unknown>>(
  object: T,
): Output<T> =>
  Object.keys(object).reduce(
    (acc, key) => {
      const value = object[key]
      if (value === undefined || value === null) return acc

      // TODO: handle nested objects
      acc[key] = object[key]
      return acc
    },
    {} as Record<RecordKeyType, unknown>,
  ) as Output<T>
