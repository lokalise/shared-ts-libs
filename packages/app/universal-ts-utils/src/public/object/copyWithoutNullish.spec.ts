import { describe, expect, it } from 'vitest'
import { copyWithoutNullish } from './copyWithoutNullish.js'

describe('copyWithoutNullish', () => {
  it('Does nothing when there are no undefined fields', () => {
    const result = copyWithoutNullish({
      a: 'a',
      b: '',
      c: ' ',
      d: null,
      e: {},
    })

    expect(result).toMatchInlineSnapshot(`
      {
        "a": "a",
        "b": "",
        "c": " ",
        "e": {},
      }
    `)
  })

  it('Removes undefined fields', () => {
    const result = copyWithoutNullish({
      a: undefined,
      b: 'a',
      c: '',
      d: undefined,
      e: ' ',
      f: null,
      g: {
        someParam: 12,
      },
      h: undefined,
    })

    const varWithNarrowedType = result satisfies Record<
      string,
      string | Record<string, unknown> | null
    >
    const bValue: string = varWithNarrowedType.b
    const gValue: {
      someParam: number
    } = varWithNarrowedType.g

    expect(bValue).toBe('a')
    expect(gValue).toEqual({
      someParam: 12,
    })

    expect(result).toMatchInlineSnapshot(`
      {
        "b": "a",
        "c": "",
        "e": " ",
        "g": {
          "someParam": 12,
        },
      }
    `)
  })
})
