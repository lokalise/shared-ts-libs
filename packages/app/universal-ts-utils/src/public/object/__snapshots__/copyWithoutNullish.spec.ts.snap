// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`copyWithoutNullish > Does nothing when there are no undefined fields 1`] = `
{
  "a": "a",
  "b": "",
  "c": " ",
  "e": {},
}
`;
